examples/plot_adversarial_mitigation.py

```diff
@@ -0,0 +1,366 @@
+        torch.nn.init.xavier_normal_(m.bias.data, gain=0.1)
+
+
+predictor_model.apply(weights_init)
+adversary_model.apply(weights_init)
+
+# %%
+# We may define the optimizers however we like. In this case, let's use a
+# predictor optimizer with SGD with decaying learning rate, and Adam for the
+# adversary.
+predictor_optimizer = torch.optim.SGD(predictor_model.parameters(), lr=0.5, momentum=0.9, nesterov=True)
+adversary_optimizer = torch.optim.Adam(adversary_model.parameters(), lr=0.1)
+scheduler1 = torch.optim.lr_scheduler.LambdaLR(predictor_optimizer, lr_lambda=lambda epoch: 1/(3*epoch+2))
+
+# %%
+# Then, the instance itself. We Will define the loss function with
+# weights because we are dealing with an imbalanced dataset.
+
+mitigator = AdversarialFairnessClassifier(
+    predictor_model=predictor_model,
+    adversary_model=adversary_model,
+    predictor_loss=torch.nn.BCEWithLogitsLoss(
+        pos_weight=torch.ones([1])*mean(Y == 1.)),
+    adversary_loss=torch.nn.BCEWithLogitsLoss(
+        pos_weight=torch.ones([1])*mean(Z == 1.)),
+    predictor_function="binary",
+    optimizer=None,
+    predictor_optimizer=predictor_optimizer,
+    adversary_optimizer=adversary_optimizer,
+    alpha=0.1,
+    constraints='demographic_parity'
+)
+
+# Carefully tweaked ;)
+epochs = 13
+batch_size = 2**8
+
+# %%
+# Instead of only looking at training loss, we also take a look at some validation
+# metrics. For this, we chose the demographic parity difference to check to what
+# extent the constraint (demographic parity in this case) is satisfied.
+
+
+def validate():
+    predictions = mitigator.predict(X_test)
+    dp_diff = demographic_parity_difference(Y_test,
+                                            predictions,
+                                            sensitive_features=Z_test)
+    accuracy = mean(predictions == Y_test)
+    selection_rate = mean(predictions == 1.)
+    print("DP diff: {:.4f}, accuracy: {:.4f}, selection_rate: {:.4f}".format(
+        dp_diff, accuracy, selection_rate))
+
+# %%
+# We make use of a callback function to incorporate the validation into the training
+# loop. Additionally, we handle the prediction_optimizer updates here.
+
+
+epoch_count = 1
+
+
+def callback_fn():
+    global epoch_count
+    mitigator.alpha = 0.1 * sqrt(epoch_count)
+    epoch_count += 1
+    scheduler1.step()
+
+    validate()
+
+# %%
+# Finally, we fit the model
+
+
+mitigator.fit(
+    X,
+    Y,
+    sensitive_features=Z,
+    epochs=epochs,
+    batch_size=batch_size,
+    shuffle=True,
+    progress_updates=None,
+    callback_fn=callback_fn)
+
+
+# %%
+# Then, we more carefully improve this by changing the model. That is, we do not
+# change the predictor/adversary model parameters, but we do change the learning
+# parameters. In particular, we set very small learning rates and alpha a bit higher.
+# As mentioned earlier, training adversarially is tricky, and ideally
+# you'd train this in a controlled way.
```

Comment:
With "controlled" do you mean systematically go over different settings or something else?

Reply:
ah yea i could be more clear, I meant "in a slow and careful way" because training can quickly diverge and you want to avoid this.